### PR TITLE
New Beta Resource - `azurerm_function_app_function`

### PR DESCRIPTION
**Pull request by jackofallops**

### CODE REVIEW

**Comment by jackofallops** [comment]
Testing looking good

image

File: internal/services/appservice/function_app_function_resource.go [hunk 1]
```diff
@@ -0,0 +1,490 @@
...
+			Type:        pluginsdk.TypeBool,
+			Optional:    true,
+			Default:     true,
+			Description: "Should this function be enabled. Defaults to `true`.",
+		},
+
+		"language": {
+			Type:         pluginsdk.TypeString,
+			Optional:     true,
+			ValidateFunc: validation.StringIsNotEmpty, // TODO - find the valida list of strings
```
Comment (mbfrahry) [inline internal/services/appservice/function_app_function_resource.go:93]: Is this todo something we need to do before merging?

File: internal/services/appservice/function_app_function_resource.go [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+			}
+
+			appId, err := parse.FunctionAppID(appFunction.AppID)
+			if err != nil {
+				return err
+			}
+
+			id := parse.NewFunctionAppFunctionID(appId.SubscriptionId, appId.ResourceGroup, appId.SiteName, appFunction.Name)
+
+			// TODO - Do we need to (or can we?) check the function runtime is ready?
```
Comment (mbfrahry) [inline internal/services/appservice/function_app_function_resource.go:195]: Another todo to check?

**Review by mbfrahry — APPROVED** [review]
LGTM but there are a few todos that haven't been addressed. Not sure if they need to be before merging

File: internal/services/appservice/function_app_function_resource.go (hunk 2 above)
Comment (jackofallops) [reply internal/services/appservice/function_app_function_resource.go:195]: one I forgot to remove in the latest commit 🙈

**Comment by jackofallops** [comment]
Retested after review feedback changes:

image

File: internal/services/appservice/function_app_function_resource_test.go [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+  function_app_id = azurerm_function_app_function.test.function_app_id
+  language        = azurerm_function_app_function.test.language
+  test_data       = azurerm_function_app_function.test.test_data
+  config_json     = azurerm_function_app_function.test.config_json
+}
+
+`, r.basic(data))
+}
+
+func (r FunctionAppFunctionResource) template(data acceptance.TestData) string {
```
Comment (tombuildsstuff) [inline internal/services/appservice/function_app_function_resource_test.go:239]: pedantic, but as Windows is labelled below we should do the same here:

```suggestion
func (r FunctionAppFunctionResource) templateLinux(data acceptance.TestData) string {
```

File: website/docs/r/function_app_function.html.markdown [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    application_stack {
+      dotnet_version = "6"
+    }
+  }
+}
+
+resource "azurerm_function_app_function" "example" {
+  name            = "example-function-app-function"
+  function_app_id = azurerm_linux_function_app.example.id
+  language        = "Python"
```
Comment (tombuildsstuff) [inline website/docs/r/function_app_function.html.markdown:125]: this is a .net 6 example so:

```suggestion
  language        = "CSharp"
```

File: internal/services/appservice/function_app_function_resource.go [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+			if _, err = createWait.WaitForStateContext(ctx); err != nil {
+				return fmt.Errorf("waiting for %s to be ready", *appId)
+			}
+
+			locks.ByID(appId.ID())
+			defer locks.UnlockByID(appId.ID())
+
+			future, err := client.CreateFunction(ctx, id.ResourceGroup, id.SiteName, id.FunctionName, fnEnvelope)
+			if err != nil {
+				fn, _ := client.Get(ctx, id.ResourceGroup, id.SiteName)
```
Comment (tombuildsstuff) [inline internal/services/appservice/function_app_function_resource.go:262]: we should be checking this error

**Review by tombuildsstuff — APPROVED** [review]
Left a few minor comments but this otherwise LGTM 👍

**Pull request merged**